agent/pkg/registrator/registrator_test.go

```diff
@@ -0,0 +1,201 @@
+package registrator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"testing"
+
+	"k8s.io/client-go/kubernetes/fake"
+	k8stest "k8s.io/client-go/testing"
+
+	l5dk8s "github.com/linkerd/linkerd2/pkg/k8s"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	clientID     = "client-id"
+	clientSecret = "client-secret"
+	accessToken  = "eyJhbGciOiJIUzI6IkpXVCJ9.eyJzdWIiOiIxMjMNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2Q6yJV_adQssw5c"
+	newAgentID   = "new-agent-id"
+)
+
+type token struct {
+	AccessToken string `json:"access_token"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+func TestRegistrator(t *testing.T) {
+	fixtures := []*struct {
+		testName           string
+		configMap          *corev1.ConfigMap
+		expInfo            *AgentInfo
+		expErr             error
+		expectRegistration bool
+	}{
+		{
+			"manifest fully hydrated (does not perform registration)",
+			&corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      agentMetadataConfigMap,
+					Namespace: namespace,
+				},
+				Data: map[string]string{
+					agentIDKey:   "agent-id",
+					agentNameKey: "agent-name",
+				},
+			},
+			&AgentInfo{
+				AgentName:  "agent-name",
+				AgentID:    "agent-id",
+				IsNewAgent: false,
+			},
+			nil,
+			false,
+		},
+		{
+			"only name present (performs registration)",
+			&corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      agentMetadataConfigMap,
+					Namespace: namespace,
+				},
+				Data: map[string]string{
+					agentNameKey: "agent-name",
+				},
+			},
+			&AgentInfo{
+				AgentName:  "agent-name",
+				AgentID:    newAgentID,
+				IsNewAgent: true,
+			},
+			nil,
+			true,
+		},
+		{
+			"missing agent name",
+			&corev1.ConfigMap{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      agentMetadataConfigMap,
+					Namespace: namespace,
+				},
+				Data: map[string]string{
+					agentIDKey: "agent-id",
+				},
+			},
+			nil,
+			fmt.Errorf("%s config map needs to have an agent_name key", agentMetadataConfigMap),
+			false,
+		},
+		{
+			"missing config map",
+			nil,
+			nil,
+			fmt.Errorf("could not find %s config map", agentMetadataConfigMap),
+			false,
+		},
+	}
+
+	for _, tc := range fixtures {
+		tc := tc
+		t.Run(tc.testName, func(t *testing.T) {
+			objs := []runtime.Object{}
+			if tc.configMap != nil {
+				objs = append(objs, tc.configMap)
+			}
+			cs := fake.NewSimpleClientset(objs...)
+			k8sApi := &l5dk8s.KubernetesAPI{
+				Interface: cs,
+			}
+
+			tokenRequests := 0
+			registrationRequests := 0
+			apiSrv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
nice!

related to comment around creating a bcloud api client, separate from `AgentRegistrator`, we could bundle this `httptest.NewServer` as part of a mock setup.